src/listen/ldap_sync/active_directory.c

```diff
@@ -27,6 +27,297 @@
+				break;
+			}
+
+			node->value = fr_value_box_alloc_null(node);
+
+			switch (node->op) {
+			case LDAP_FILTER_OP_EQ:
+				fr_value_box_bstrndup(node, node->value, NULL, fr_sbuff_current(&value_marker),
+						      fr_sbuff_diff(sbuff, &value_marker), false);
+
+				/*
+				 *	Filter values with * in them are substring matches rather than equality
+				 */
+				if (strchr(node->value->vb_strvalue, '*')) node->op = LDAP_FILTER_OP_SUBSTR;
+				break;
+
+			/*
+			 *	Since we don't have the LDAP schema, we make an assumption that <=, >= and
+			 *	bitwise operators are going to be used with numeric attributes
+			 */
+			case LDAP_FILTER_OP_GE:
+			case LDAP_FILTER_OP_LE:
+			case LDAP_FILTER_OP_BIT_AND:
+			case LDAP_FILTER_OP_BIT_OR:
+				fr_value_box_from_str(node, node->value, FR_TYPE_UINT32, NULL,
+						      fr_sbuff_current(&value_marker),
+						      fr_sbuff_diff(sbuff, &value_marker), NULL, false);
+				break;
+
+			/*
+			 *	Operator should not be any of these at the end of a filter
+			 */
+			case LDAP_FILTER_OP_PRESENT:
+			case LDAP_FILTER_OP_SUBSTR:
+			case LDAP_FILTER_OP_UNSET:
+				fr_assert(0);
+				break;
+			}
+			/*
+			 *	Take a copy of the original filter for debug output
+			 */
+			node->orig = talloc_zero_array(node, char, fr_sbuff_diff(sbuff, &filter_marker) + 1);
+			memcpy(node->orig, fr_sbuff_current(&filter_marker), fr_sbuff_diff(sbuff, &filter_marker));
+			DEBUG3("%*sParsed LDAP filter (%s)", depth, "", node->orig);
+
+			/*
+			 *	If we're at the very top level we should be at the end
+			 *	of the buffer, otherwise we should be able to advance
+			 */
+			if (depth == 0) {
+				return (fr_sbuff_remaining(sbuff) == 0 ? 0 : -1);
+			} else {
+				return fr_sbuff_advance(sbuff, 1);
+			}
+
+		case '<':
+		case '>':
+			if (fr_sbuff_remaining(sbuff) < 2) {
+				fr_strerror_printf("Insufficient characters after \"%c\"", fr_sbuff_current(sbuff)[0]);
+				return -1;
+			}
+			if (fr_sbuff_current(sbuff)[1] != '=') {
+				fr_strerror_printf("Missing \"=\" after \"%c\"", fr_sbuff_current(sbuff)[0]);
+				return -1;
+			}
+			node->op = fr_sbuff_current(sbuff)[0] == '<' ? LDAP_FILTER_OP_LE : LDAP_FILTER_OP_GE;
+			fr_sbuff_marker(&attr_e_marker, sbuff);
+			break;
+
+		case '~':
+			fr_strerror_const("Approximate match filters not allowed");
+			return -1;
+
+		default:
+			fr_strerror_printf("\"%c\" found in unexpected place", fr_sbuff_current(sbuff)[0]);
+			return -1;
+		}
+		fr_sbuff_advance(sbuff, 1);
+	}
+	return 0;
+}
+
+/** Parse an LDAP filter into its component nodes
+ *
+ * @param[in] ctx	to allocate nodes in.
+ * @param[out] root	of tree of parsed nodes.
+ * @param[in] filter	to parse.
+ * @param[in] attrs	specified in the sync config.  Used to warn user if the filter
+ *			references attributes not being requested.
+ * @return
+ *	- 0 on success
+ *	< 0 on failure
+ */
+static int ldap_filter_parse(TALLOC_CTX *ctx, ldap_filter_t **root, char const *filter, char const **attrs)
```

Comment:
If this is generic to LDAP and not to AD, it might be better to put this code into `src/lib/ldap`, so that other things can use it.

But that's a minor nit.